### PR TITLE
enhancement/3 Filter optimization

### PR DESCRIPTION
**Pull request by nikita-parkhomenko**
- Store api queries and data to users state
- Use one FILTER_USERS action type for all filter operations
- Optimize filter users saga
- Replaced search input from redux to react form
- Handle input events
- Reuse action button

### CODE REVIEW

File: src/assets/icons.js [hunk 1]
```diff
@@ -43,3 +43,23 @@ export const plus = (<svg width="1em"
     <path fillRule="evenodd"
           d="M8 4a.5.5 0 0 1 .5.5v3h3a.5.5 0 0 1 0 1h-3v3a.5.5 0 0 1-1 0v-3h-3a.5.5 0 0 1 0-1h3v-3A.5.5 0 0 1 8 4z"/>
 </svg>);
+
+export const editIcon = (<svg width="1em"
```
Comment (ichabanivan) [inline src/assets/icons.js:47]: fontawesome

File: src/components/custom-button.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import { sortDown, sortUp } from '../assets/icons';
+import { selector as usersReducer } from '../pages/private/users/reducer';
+
+const CustomButton = ({ sortUsers, filterName }) => {
+    const { sortASC, sort } = useSelector(usersReducer);
+    return (
+        <Button
+            className="text-info py-0"
+            color="link"
+            onClick={event => sortUsers(event)}
```
Comment (ichabanivan) [inline src/components/custom-button.js:14]: memoize

File: src/components/custom-button.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        >
+            { sort === filterName.toLowerCase() && sortASC ? sortDown : sortUp }
```
Comment (ichabanivan) [inline src/components/custom-button.js:16]:
```suggestion
            { sort === filterName && sortASC ? false : true }
```
so?

File: src/pages/private/create-user/create-user.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+
+// outsource dependencies
+import React from 'react';
+import { Container, Row, Col } from 'reactstrap';
+
+const CreateUser = () => {
+    return (
```
Comment (ichabanivan) [inline src/pages/private/create-user/create-user.js:7]: why return?

File: src/pages/private/create-user/saga.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+// outsource dependencies
+import { takeEvery, put, delay } from 'redux-saga/effects';
+
+// local dependencies
+import { TYPE } from './reducer';
+
+function * initializeSaga() {
+    yield delay(1000);
+    try {
+        yield put({ type: TYPE.META, payload: { initialized: true }});
```
Comment (ichabanivan) [inline src/pages/private/create-user/saga.js:11]: outside try catch

File: src/pages/private/users/saga.js [hunk 6]
```diff
@@ -27,56 +27,28 @@ function * getUsersSaga() {
...
-        const searchedUsers = yield call(ApiService.findUser, searchUser);
-        yield put({ type: TYPE.META, payload: { users: searchedUsers.data.content } });
+        yield put({ type: TYPE.META, payload });
+        let { name, page, sort, size, sortASC } = yield select(usersSelector);
+        sort = `${sort},${sortASC ? 'ASC' : 'DESC'}`;
+
+        const users = yield call(
+            ApiService.filterUsers,
+            { name },
+            { page, sort, size }
```
Comment (ichabanivan) [inline src/pages/private/users/saga.js:41]:
```suggestion
            { page, sort: `${sort},${sortASC ? 'ASC' : 'DESC'}`, size }
```

File: src/pages/private/users/saga.js [hunk 7]
```diff
@@ -27,56 +27,28 @@ function * getUsersSaga() {
...
-    yield delay(1000);
+    yield delay(1000 / 2);
     try {
-        const searchedUsers = yield call(ApiService.findUser, searchUser);
-        yield put({ type: TYPE.META, payload: { users: searchedUsers.data.content } });
+        yield put({ type: TYPE.META, payload });
+        let { name, page, sort, size, sortASC } = yield select(usersSelector);
+        sort = `${sort},${sortASC ? 'ASC' : 'DESC'}`;
+
+        const users = yield call(
```
Comment (zanusilker) [inline src/pages/private/users/saga.js:38]:
```suggestion
        const{ data, totalElements ...} = yield call(
```

File: src/pages/private/users/saga.js [hunk 8]
```diff
@@ -27,56 +27,28 @@ function * getUsersSaga() {
...
+function * filterUsersSaga({ payload }) {
     yield put({ type: TYPE.META, payload: { disabled: true } });
-    yield delay(1000);
+    yield delay(1000 / 2);
     try {
-        const searchedUsers = yield call(ApiService.findUser, searchUser);
-        yield put({ type: TYPE.META, payload: { users: searchedUsers.data.content } });
+        yield put({ type: TYPE.META, payload });
+        let { name, page, sort, size, sortASC } = yield select(usersSelector);
+        sort = `${sort},${sortASC ? 'ASC' : 'DESC'}`;
```
Comment (ichabanivan) [inline src/pages/private/users/saga.js:36]: no!

File: src/pages/private/users/users.js [hunk 9]
```diff
@@ -1,53 +1,54 @@
...
+            page,
+            name,
             totalPages,
     } = useSelector(usersSelector);
 
-    const sortUsers = useCallback(event => {
-        const sortBy = event.target.textContent;
-        dispatch({ type: TYPE.SORT_USERS, payload: { sortBy } });
+    const searchByName = useCallback(event => {
+        event.preventDefault();
```
Comment (ichabanivan) [inline src/pages/private/users/users.js:32]: BUTTON TYPE BUTTON

File: src/pages/private/users/users.js [hunk 10]
```diff
@@ -107,32 +119,24 @@ const Users = ({ handleSubmit, pristine, reset, submitting }) => {
...
-                        <Button
-                            onClick={e => sortUsers(e)}
-                            className="text-info py-0"
-                            color="link"
-                        >
-                            { sortNamesASC ? sortDown : sortUp }
-                            <span className="font-weight-bold">Name</span>
-                        </Button>
+                        <CustomButton
+                            sortUsers={sortUsers}
```
Comment (ichabanivan) [inline src/pages/private/users/users.js:123]:
```suggestion
                            sortUsers={sortUsers('name)}
```

File: src/pages/private/users/users.js [hunk 11]
```diff
@@ -1,53 +1,54 @@
...
 
-    const nextPage = useCallback(pageNumber => {
-        dispatch({ type: TYPE.NEXT_PAGE, payload: { pageNumber } });
+    const sortUsers = useCallback(event => {
+        const sort = event.target.textContent.toLowerCase();
+        dispatch({ type: TYPE.FILTER_USERS, payload: { sort, sortASC: !sortASC } });
+    }, [dispatch, sortASC]);
+
+    const changeSearchFieldHandler = useCallback(event => {
+        const name = event.target.value.trimLeft();
```
Comment (ichabanivan) [inline src/pages/private/users/users.js:46]: -

File: src/pages/private/users/users.js [hunk 12]
```diff
@@ -70,35 +71,46 @@ const Users = ({ handleSubmit, pristine, reset, submitting }) => {
...
+                                    </Button>
+                                }
+                            </InputGroupAddon>
+                            <Input
+                                placeholder="Search"
+                                type="text"
+                                disabled={disabled}
+                                value={name}
+                                name='searchUser'
+                                onChange={event => changeSearchFieldHandler(event)}
```
Comment (ichabanivan) [inline src/pages/private/users/users.js:95]:
```suggestion
                                onChange={event => changeSearchFieldHandler(event.target.value)}
```

Comment (ichabanivan) [reply src/pages/private/users/users.js:95]: memoize

File: src/pages/private/users/users.js [hunk 13]
```diff
@@ -107,32 +119,24 @@ const Users = ({ handleSubmit, pristine, reset, submitting }) => {
...
                     </th>
                     <th>Role</th>
                     <th>Actions</th>
                 </tr>
                 </thead>
                 <tbody>
                 {
-                    users.length &&
+                    users.length !== 0 ?
                         users.map((user) => (
```
Comment (ichabanivan) [inline src/pages/private/users/users.js:140]:
```suggestion
                       (users || []).map((user) => (
```

**Pull request merged**